Answer response refactor

Dokotela opened the pull request:
This is more in case you want to review, not really ready for pull yet.

I'm running into some issues because I added a group class (single survey holds multiple groups) instead of multiple surveys that function like groups.

Also, because we're allowing multiple answers, I often had to switch to answers[0] ".value.responseType.value" -> ".value.answers[0].value"

Changed most of the codes to linkId (or surveyLinkId, groupLinkId)

Having issues with ItemGroup, Question and SurveyItems. I'm also not sure I have all your mapping functions working correctly.

Dokotela committed:
Fixed commands. QuestionLinkId & surveyLinkId. Used answers[0]

Dokotela committed:
Fixed answer_items, changed "code" -> "linkId"

Dokotela committed:
Few little changes to state linkId

Dokotela committed:
Changed ".value.responseType.value" -> ".value.answers[0].value"

Dokotela committed:
Changed lots of surveys. to surveys.surveyItems

Dokotela committed:
Changed survey to group in most places

Dokotela committed:
Issues as expected with ItemGroup, Question and SurveyItems

FireJuun reviewed (commented):
Looking at this briefly... The changes to PRAPARE's JSON calls seem good.

If Answers is still a set, you can't use [0] to call it. That only works on lists. Instead, Answers.elementAt(index) will work with either, though for a list it's probably overkill.

Better yet, just pass Answer forward. This will provide the reference to the relevant answer class (rather than cloning it), and you'll have a lot fewer errors in the long run than dealing with set/list indexes.

Will keep skimming / reviewing...

FireJuun commented on lib/controllers/fhir_questionnaire/questionnaire_controller.dart:
```diff
@@ -13,52 +12,49 @@ class QuestionnaireController extends GetxController {
+                    // if AnswerType.choice ==> use booleans, otherwise strings
+                    // todo: update with various additional response types
+                    (ans.code == 'decimal' || ans.code == 'string')
+                        ? UserResponse(
```
This can be a switch statement now

FireJuun replied:
If we want to add different answer types, we can do that here in a switch statement.

FireJuun commented on lib/controllers/fhir_questionnaire/questionnaire_controller.dart:
```diff
@@ -67,18 +63,15 @@ class QuestionnaireController extends GetxController {
+              (q as Question).linkId,
+              UserResponse(
+                  questionLinkId: q.linkId,
+                  answers: [AnswerBoolean(false)]).obs,
```
What does the initial boolean represent now after the refactor? 
Initially, I had a UserResponse created for every possible question that defaulted to false and would change to true when answered/checked. Is that still the case?

FireJuun replied:
Answer: this is the single function that decides if an answer has been given (used in validation). In true/false questions, it also lists the true/false boolean.

Dokotela replied:
I think you and I are using AnswerBooleans differently. I was only using an AnswerBoolean to record an answer to a boolean question. It looks to me like you're using AnswerBoolean as a marker to say that answer has been selected. If you wanted to keep that, I could add a boolean field to UserResponse called selected, to keep track of which one has been selected. Also, I was making a list of answers to allow for more than one answer to a question. But (if I understand correctly), you're creating a UserResponse for each possible answer. If that's the case, would it make more sense to have a single answer for each userResponse, and then you could pass a list of UserResponses back to me? I can match up the questions and answers using the linkIds, and I'll know if that question is allowed more than one response.

FireJuun reviewed (commented):
Here are the other comments I had on first / superficial review.

I believe your List<Answer> called Answers will always assume that the index 0 value is the boolean as to whether an answer exists or not. Assuming the index 0 value is never changed or removed, that could work. If that's the case, we'll need to be very careful whenever we add/remove elements from that list

Dokotela committed:
Changed _MapAllQuestions to work with groups, used switch

Dokotela committed:
Commented out _mapAllActiveResponses

Dokotela commented:
Think I fixed _MapAllQuestions, not sure what to do with _mapAllActiveResponses

FireJuun committed:
update ios

FireJuun committed:
Merge branch 'answer_response_refactor' of https://github.com/Dokotela/prapare into answer_response_refactor

FireJuun committed:
fix: incorrect casting of ItemGroups and Questions
Previously, these were causing errors in data mapping
Teporarily turned off commands, as they need to be handled separately

FireJuun commented:
This refactor is getting a lot closer. I had to get rid of the automatic casting from prior commits, since it was causing errors.

Running in debug mode, it seems to be correctly calling the FHIR classes (including subquestions) from your refactor. The next steps are to update the UI and the controllers to correctly parse said FHIR classes and to handle subquestions.

FireJuun committed:
fix: correctly handle parsing of question titles

FireJuun committed:
fix: question/subquestion titles work correctly 🎉

FireJuun committed:
fix: missing PrapareCodesUtil value + code cleanup

Dokotela committed:
Added saveResponse() method to QuestionnaireModel

Dokotela committed:
Merge branch 'answer_response_refactor' of https://github.com/Dokotela/prapare into answer_response_refactor

Dokotela commented:
Added saveResponse() method to QuestionnaireModel. Should now save locally.

FireJuun committed:
feat: Answer AnswerResponse now self-aware of type
This means they no longer need to run a util function on each UI build
This also removes some opportunities for failure if data types change
Finally, this makes it easier to build the UI w/ self-aware data classes

FireJuun committed:
Merge branch 'answer_response_refactor' of https://github.com/Dokotela/prapare into answer_response_refactor

FireJuun committed:
ios pod updates due to sqflite package integration

FireJuun committed:
fix: AnswerResponse itemTypes  work as intended

FireJuun committed:
add 'choose not 2 answer' to ?s with type:string

FireJuun committed:
code cleanup

FireJuun committed:
refact: extract all validation into one controller

FireJuun committed:
code cleanup. change survey -> group 4 consistency

FireJuun committed:
temp disabled radio/✓boxes to handle string inputs

FireJuun committed:
fix: radio/check/string/num items now display
for number items, we need to handle null values
added util to compare ItemType and QuestionnaireItemType

FireJuun committed:
feat: UserResponse classes have distinct ItemTypes

FireJuun committed:
breaking: completely reorganized question UI
radio button group values are held in stateful widget instead of controller
this makes it much easier to manage when nesting your UI

FireJuun committed:
refactor: item / items for naming consistency

FireJuun committed:
fix: checkbox items now appropriately toggle

FireJuun committed:
re-implement views for nested questions

FireJuun committed:
remove cached files (.gitignore fix)

FireJuun committed:
re-add cached files (.gitignore fix)

FireJuun committed:
update gitignore

FireJuun committed:
remove cached files (.gitignore fix)

FireJuun committed:
update gitignore

FireJuun committed:
re-add cached files (.gitignore fix)
